gpflow/features.py

```diff
@@ -0,0 +1,184 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.from __future__ import print_function
+
+from abc import ABCMeta, abstractmethod
+from functools import singledispatch
+
+import numpy as np
+import tensorflow as tf
+
+from . import conditionals, transforms, kernels, decors, settings
+from .params import Parameter, Parameterized
+
+
+class InducingFeature(Parameterized, metaclass=ABCMeta):
+    """
+    Abstract base class for inducing features.
+    """
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Returns the number of features, relevant for example to determine the
+        size of the variational distribution.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def Kuu(self, kern, jitter=0.0):
+        """
+        Calculates the covariance matrix between features for kernel `kern`.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def Kuf(self, kern, Xnew):
+        """
+        Calculates the covariance matrix with function values at new points
+        `Xnew` for kernel `kern`.
+        """
+        raise NotImplementedError()
+
+
+class InducingPoints(InducingFeature):
+    """
+    Real-space inducing points
+    """
+
+    def __init__(self, Z):
+        """
+        :param Z: the initial positions of the inducing points, size M x D
+        """
+        super().__init__()
+        self.Z = Parameter(Z)
+
+    def __len__(self):
+        return self.Z.shape[0]
+
+    @decors.params_as_tensors
+    def Kuu(self, kern, jitter=0.0):
+        Kzz = kern.K(self.Z)
+        Kzz += jitter * tf.eye(len(self), dtype=settings.dtypes.float_type)
+        return Kzz
+
+    @decors.params_as_tensors
+    def Kuf(self, kern, Xnew):
+        Kzx = kern.K(self.Z, Xnew)
+        return Kzx
+
+
+class Multiscale(InducingPoints):
+    """
+    Multi-scale inducing features
+    Originally proposed in
+
+    ::
+
+      @incollection{NIPS2009_3876,
+        title = {Inter-domain Gaussian Processes for Sparse Inference using Inducing Features},
+        author = {Miguel L\'{a}zaro-Gredilla and An\'{\i}bal Figueiras-Vidal},
+        booktitle = {Advances in Neural Information Processing Systems 22},
+        year = {2009},
+      }
+
+    """
+
+    def __init__(self, Z, scales):
+        super().__init__(Z)
+        self.scales = Parameter(scales,
+                                transform=transforms.positive)  # Multi-scale feature widths (std. dev. of Gaussian)
+        assert self.Z.shape == scales.shape
+
+    def _cust_square_dist(self, A, B, sc):
+        """
+        Custom version of _square_dist that allows sc to provide per-datapoint length
+        scales. sc: N x M x D.
+        """
+        return tf.reduce_sum(tf.square((tf.expand_dims(A, 1) - tf.expand_dims(B, 0)) / sc), 2)
+
+    @decors.params_as_tensors
+    def Kuf(self, kern, Xnew):
+        if isinstance(kern, kernels.RBF):
```

Comment:
```
if not isinstance(kern, kernels.RBF):
    raise NotImplementedError(...)
....
```

Reply:
My thinking was that other kernels may be implemented in the future.